Source/WebKit/UIProcess/Gamepad/wpe/GamepadProviderWPE.cpp

```diff
@@ -0,0 +1,181 @@
+{
+    unsigned index = 0;
+    while (index < m_gamepadVector.size() && m_gamepadVector[index])
+        ++index;
+
+    auto device = makeUnique<PlatformGamepadWPE>(gamepad, index);
+    if (m_gamepadVector.size() <= index)
+        m_gamepadVector.grow(index + 1);
+
+    m_gamepadVector[index] = device.get();
+    m_gamepadMap.add(gamepad, WTFMove(device));
+
+    if (m_isMonitoringInput)
+        wpe_gamepad_start_input_monitor(gamepad);
+
+    auto eventVisibility = isInitialDevice == IsInitialDevice::No ? EventMakesGamepadsVisible::Yes : EventMakesGamepadsVisible::No;
+    for (auto& client : m_clients)
+        client.platformGamepadConnected(*m_gamepadVector[index], eventVisibility);
+}
+
+void GamepadProviderWPE::gamepadDisconnected(WPEGamepad* gamepad)
+{
+    auto device = m_gamepadMap.take(gamepad);
+    if (!device)
+        return;
+
+    auto index = m_gamepadVector.find(device.get());
+    if (index != notFound)
+        m_gamepadVector[index] = nullptr;
+
+    for (auto& client : m_clients)
+        client.platformGamepadDisconnected(*device);
+}
+
+void GamepadProviderWPE::startMonitoringGamepads(GamepadProviderClient& client)
+{
+    bool shouldCreateManager = m_clients.isEmptyIgnoringNullReferences();
+
+    ASSERT(!m_clients.contains(client));
+    m_clients.add(client);
+
+    if (!shouldCreateManager)
+        return;
+
+    ASSERT(m_gamepadVector.isEmpty());
+
+    m_manager = adoptGRef(wpe_display_create_gamepad_manager(wpe_display_get_primary()));
+    if (!m_manager)
+        return;
+
+    g_signal_connect_swapped(m_manager.get(), "device-added", G_CALLBACK(+[](GamepadProviderWPE* provider, WPEGamepad* gamepad) {
+        provider->gamepadConnected(gamepad, IsInitialDevice::No);
+    }), this);
+    g_signal_connect_swapped(m_manager.get(), "device-removed", G_CALLBACK(+[](GamepadProviderWPE* provider, WPEGamepad* gamepad) {
+        provider->gamepadDisconnected(gamepad);
+    }), this);
+
+    m_isMonitoringInput = true;
+    gsize deviceCount;
+    GUniquePtr<WPEGamepad*> gamepads(wpe_gamepad_manager_list_devices(m_manager.get(), &deviceCount));
+    for (size_t i = 0; i < deviceCount; ++i)
+        gamepadConnected(gamepads.get()[i], IsInitialDevice::Yes);
+}
+
+void GamepadProviderWPE::stopMonitoringGamepads(GamepadProviderClient& client)
+{
+    ASSERT(m_clients.contains(client));
+
+    bool shouldDestroyManager = m_clients.remove(client) && m_clients.isEmptyIgnoringNullReferences();
+    if (shouldDestroyManager) {
+        if (m_manager) {
+            g_signal_handlers_disconnect_by_data(m_manager.get(), this);
+            m_manager = nullptr;
+        }
+        m_gamepadVector.clear();
+        m_gamepadMap.clear();
+        m_isMonitoringInput = false;
+    }
+}
+
+void GamepadProviderWPE::playEffect(unsigned, const String&, GamepadHapticEffectType, const GamepadEffectParameters&, CompletionHandler<void(bool)>&& completionHandler)
+{
+    // Not supported by this provider.
```

Comment:
I think it would make sense to add a `notImplemented()` marker here, because we could use [manette_device_rumble()](https://gnome.pages.gitlab.gnome.org/libmanette/doc/0.2/method.Device.rumble.html) in the future to complete the implementation. 

```suggestion
    // Not supported by this provider.
    notImplemented();
```